File: Device/ServoCalibration.cs
// ServoCalibration.cs
// Handles servo PWM control and calibration logic
using LabJack;

namespace Device
{
    public class ServoCalibration
    {
        private LabJackDevice device;
        private int coreFrequency;
        private int pwmDIO;
        private int clockDivisor = 1;
        private int desiredFrequency;
        
        public ServoCalibration(LabJackDevice device, int coreFrequency, int pwmDIO, int clockDivisor = 1, int desiredFrequency = 50)
        {
            this.device = device;
            this.coreFrequency = coreFrequency;
            this.pwmDIO = pwmDIO;
            this.clockDivisor = clockDivisor;
            this.desiredFrequency = desiredFrequency;
        }



        public float CalculateDutyCycle(float angle)
        {
            // 0 deg = 1ms, 180 deg = 2ms, 50Hz (20ms period)
            // duty_cycle = 5 + (angle / 36)
            // Determined using common servo control data
            return 5 + (angle / 36);
        }

        public void SetServoAngle(float angle)
        {
            try
            {
                Console.WriteLine(angle);
                float dutyCycle = CalculateDutyCycle(angle);
                Console.WriteLine(dutyCycle);
                int clockTickRate = coreFrequency / clockDivisor;
                int clockRollValue = clockTickRate / desiredFrequency;
                
                 // --- Calculate PWM Values ---
                // Calculate the clock tick value where the line will transition from high to low based on user defined duty cycle percentage, rounded to the nearest integer.
                int pwmConfigA = (int)(clockRollValue * ((double)dutyCycle / 100));

                // --- Configure and write values to connected device ---
                // Configure Clock Registers, use 32-bit Clock0 for this example.
                LJM.eWriteName((Int32)device.Handle, "DIO_EF_CLOCK0_DIVISOR", (double)clockDivisor);   // Set Clock Divisor.
                LJM.eWriteName(device.Handle, "DIO_EF_CLOCK0_ROLL_VALUE", clockRollValue); // Set calculated Clock Roll Value.

                // Configure PWM Registers
                LJM.eWriteName(device.Handle, String.Format("DIO{0}_EF_INDEX", pwmDIO), 0);              // Set DIO#_EF_INDEX to 0 - PWM Out.
                LJM.eWriteName(device.Handle, String.Format("DIO{0}_EF_CLOCK_SOURCE", pwmDIO), 0);       // Set DIO#_EF to use clock 0. Formerly DIO#_EF_OPTIONS, you may need to switch to this name on older LJM versions.
                LJM.eWriteName(device.Handle, String.Format("DIO{0}_EF_CONFIG_A", pwmDIO), pwmConfigA);  // Set DIO#_EF_CONFIG_A to the calculated value.
                LJM.eWriteName(device.Handle, String.Format("DIO{0}_EF_ENABLE", pwmDIO), 1);             // Enable the DIO#_EF Mode, PWM signal will not start until DIO_EF and CLOCK are enabled.

                // Configure High-Speed Counter Registers
                LJM.eWriteName(device.Handle, String.Format("DIO{0}_EF_INDEX", device.counterDIO), 7);          // Set DIO#_EF_INDEX to 7 - High-Speed Counter.
                LJM.eWriteName(device.Handle, String.Format("DIO{0}_EF_ENABLE", device.counterDIO), 1);         // Enable the High-Speed Counter.

                LJM.eWriteName(device.Handle, "DIO_EF_CLOCK0_ENABLE", 1);   // Enable Clock0, this will start the PWM signal.

            }
            catch (LJM.LJMException e)
            {
                device.showErrorMessage(e);
                device.Dispose();
                Console.WriteLine("Error occurred. Press enter to exit.");
                Console.ReadLine();
                Environment.Exit(-1);
            }
        }

        public void TurnOffPWM()
        {
            // Turn off PWM output and counter
                string[] aNames = 
                {
                    "DIO_EF_CLOCK0_ENABLE",
                    String.Format("DIO{0}_EF_ENABLE", pwmDIO),
                    String.Format("DIO{0}_EF_ENABLE", device.counterDIO),
                };

                double[] aValues = new double[device.inputPins.Length]; 
                int numFrames = aNames.Length;
            int errorAddress = -1;
                LJM.eWriteNames(device.Handle, numFrames, aNames, aValues, ref errorAddress);
        }
    }
}

File: calibrate_py/servo_calibration.py
"""Servo PWM control translated from ServoCalibration.cs."""

from typing import Optional
import math
from labjack.ljm import ljm


class ServoCalibration:
    def __init__(
        self,
        device,
        core_frequency: int,
        pwm_dio: int,
        clock_divisor: int = 1,
        desired_frequency: int = 50,
    ):
        self.device = device
        self.core_frequency = core_frequency
        self.pwm_dio = pwm_dio
        self.clock_divisor = clock_divisor
        self.desired_frequency = desired_frequency

    def calculate_duty_cycle(self, angle: float) -> float:
        return 5.0 + (angle / 36.0)

    def set_servo_angle(self, angle: float):
        try:
            duty_cycle = self.calculate_duty_cycle(angle)
            clock_tick_rate = int(self.core_frequency / self.clock_divisor)
            clock_roll_value = int(clock_tick_rate / self.desired_frequency)
            pwm_config_a = int(round(clock_roll_value * (duty_cycle / 100.0)))

            # Write clock and PWM configuration
            ljm.eWriteName(
                self.device.handle, "DIO_EF_CLOCK0_DIVISOR", float(self.clock_divisor)
            )
            ljm.eWriteName(
                self.device.handle, "DIO_EF_CLOCK0_ROLL_VALUE", float(clock_roll_value)
            )

            ljm.eWriteName(self.device.handle, f"DIO{self.pwm_dio}_EF_ENABLE", 0.0)
            ljm.eWriteName(self.device.handle, f"DIO{self.pwm_dio}_EF_INDEX", 0.0)
            ljm.eWriteName(
                self.device.handle, f"DIO{self.pwm_dio}_EF_CLOCK_SOURCE", 0.0
            )
            ljm.eWriteName(
                self.device.handle,
                f"DIO{self.pwm_dio}_EF_CONFIG_A",
                float(pwm_config_a),
            )
            ljm.eWriteName(self.device.handle, f"DIO{self.pwm_dio}_EF_ENABLE", 1.0)

            # Configure counter
            ljm.eWriteName(
                self.device.handle, f"DIO{self.device.counter_dio}_EF_ENABLE", 0.0
            )
            ljm.eWriteName(
                self.device.handle, f"DIO{self.device.counter_dio}_EF_INDEX", 7.0
            )
            ljm.eWriteName(
                self.device.handle, f"DIO{self.device.counter_dio}_EF_ENABLE", 1.0
            )

            ljm.eWriteName(self.device.handle, "DIO_EF_CLOCK0_ENABLE", 1.0)
        except Exception as e:
            self.device.show_error_message(e)
            self.device.close()
            raise

    def turn_off_pwm(self):
        names = [
            "DIO_EF_CLOCK0_ENABLE",
            f"DIO{self.pwm_dio}_EF_ENABLE",
            f"DIO{self.device.counter_dio}_EF_ENABLE",
        ]
        values = [0.0, 0.0, 0.0]
        ljm.eWriteNames(self.device.handle, len(names), names, values)
